File: frontend/src/services/RequestService.ts
// frontend/src/services/RequestService.ts

import { supabase } from '../lib/supabase';
import { v4 as uuidv4 } from 'uuid';
import type { PaymentDetails } from '../types/payment';

export type RequestType = 
  | 'team_transfer'
  | 'roster_change'
  | 'tournament_registration'
  | 'league_registration'
  | 'team_rebrand'
  | 'online_id_change'
  | 'team_creation';

export interface RequestBase {
  request_id?: string;
  request_type: RequestType;
  team_id: string;
  requested_by: string;
  requires_payment: boolean;
  metadata?: Record<string, any>;
}

export interface TeamTransferRequest extends RequestBase {
  request_type: 'team_transfer';
  new_captain_id: string;
  old_captain_id: string;
  item_id: string;
  payment_data?: PaymentDetails;
}

export interface RosterChangeRequest extends RequestBase {
  request_type: 'roster_change';
  player_id: string;
  new_role?: string;
  operation: 'add' | 'remove' | 'update';
  payment_data?: any;
}

export interface TournamentRegistrationRequest extends RequestBase {
  request_type: 'tournament_registration';
  tournament_id: string;
  player_ids: string[];
  payment_data?: any;
}

export interface LeagueRegistrationRequest extends RequestBase {
  request_type: 'league_registration';
  league_id: string;
  season?: number;
  player_ids: string[];
  payment_data?: any;
}

export interface TeamRebrandRequest extends RequestBase {
  request_type: 'team_rebrand';
  old_name: string;
  new_name: string;
  payment_data?: {
    amount: number;
    metadata: {
      request_id: string;
      request_type: string;
      team_id: string;
      old_name: string;
      new_name: string;
      logo_url?: string;
    }
  };
}

export interface OnlineIdChangeRequest extends RequestBase {
  request_type: 'online_id_change';
  player_id: string;
  old_online_id?: string;
  new_online_id: string;
  platform: string;
  payment_data?: any;
}

export interface TeamCreationRequest extends RequestBase {
  request_type: 'team_creation';
  team_name: string;
  captain_id: string;
  league_id?: string;
  payment_data?: any;
}

export type RequestData = 
  | TeamTransferRequest
  | RosterChangeRequest
  | TournamentRegistrationRequest
  | LeagueRegistrationRequest
  | TeamRebrandRequest
  | OnlineIdChangeRequest
  | TeamCreationRequest;

export class RequestService {
  private apiUrl = '/api';
  private defaultItemIds = {
    team_transfer: '1002',
    team_rebrand: '1003',
    // Add other default item IDs here
  };

  // Request builders
  public createTeamTransferRequest(params: {
    teamId: string;
    userId: string;
    newCaptainId: string;
    teamName: string;
    newCaptainName: string;
    amount: number;
  }): TeamTransferRequest {
    const requestId = uuidv4();
    const itemId = this.defaultItemIds.team_transfer;

    const paymentDetails: PaymentDetails = {
      id: uuidv4(),
      type: 'team_transfer',
      name: 'Team Ownership Transfer',
      amount: params.amount,
      description: `Transfer ownership of ${params.teamName} to ${params.newCaptainName}`,
      teamId: params.teamId,
      captainId: params.userId,
      playersIds: [],
      playerId: params.newCaptainId,
      request_id: requestId,
      referenceId: `${itemId}-${requestId.replace(/-/g, '')}`,
      item_id: itemId,
      metadata: {
        requestType: 'team_transfer',
        oldCaptainId: params.userId,
        oldCaptainName: params.userId,
        newCaptainId: params.newCaptainId,
        newCaptainName: params.newCaptainName,
        teamName: params.teamName,
        requestId: requestId,
        changeRequestData: {
          teamId: params.teamId,
          requestedBy: params.userId,
          itemId: itemId,
          playerId: params.newCaptainId,
          oldValue: params.userId,
          newValue: params.newCaptainId,
          requestId: requestId,
          metadata: {
            oldCaptainName: params.userId,
            newCaptainName: params.newCaptainName,
            teamName: params.teamName,
            requestId: requestId
          }
        }
      }
    };

    return {
      request_id: requestId,
      request_type: 'team_transfer',
      team_id: params.teamId,
      requested_by: params.userId,
      requires_payment: true,
      new_captain_id: params.newCaptainId,
      old_captain_id: params.userId,
      item_id: itemId,
      payment_data: paymentDetails
    };
  }

  // Validation methods
  private validateTeamTransferRequest(request: TeamTransferRequest): void {
    if (!request.team_id) throw new Error('Missing required field: team_id');
    if (!request.requested_by) throw new Error('Missing required field: requested_by');
    if (!request.new_captain_id) throw new Error('Missing required field: new_captain_id');
    if (!request.old_captain_id) throw new Error('Missing required field: old_captain_id');
    if (!request.item_id) throw new Error('Missing required field: item_id');

    if (request.requires_payment && !request.payment_data) {
      throw new Error('Payment data required for paid requests');
    }
  }

  private validateRequest(request: RequestData): void {
    switch (request.request_type) {
      case 'team_transfer':
        this.validateTeamTransferRequest(request);
        break;
      // Add other request type validations here
    }
  }

  // Process request with error handling and status tracking
  async processRequest(requestData: RequestData): Promise<any> {
    try {
      // Ensure request has an ID
      if (!requestData.request_id) {
        requestData.request_id = uuidv4();
      }
      
      console.log(`Processing ${requestData.request_type} request with ID: ${requestData.request_id}`);
      
      // Validate request data
      this.validateRequest(requestData);
      
      // Map request type to endpoint
      const endpoints = {
        'team_transfer': '/team/transfer',
        'roster_change': '/team/roster',
        'tournament_registration': '/tournament/register',
        'league_registration': '/league/register',
        'team_rebrand': '/team/rebrand',
        'online_id_change': '/team/online-id',
        'team_creation': '/team/create'
      };
      
      const endpoint = endpoints[requestData.request_type];
      
      if (!endpoint) {
        throw new Error(`Unknown request type: ${requestData.request_type}`);
      }
      
      // Track request status
      await this.updateRequestStatus(requestData.request_id, 'processing');
      
      // Call the API endpoint
      const response = await fetch(`${this.apiUrl}${endpoint}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(requestData)
      });
      
      if (!response.ok) {
        const errorData = await response.json();
        await this.updateRequestStatus(requestData.request_id, 'failed', errorData);
        throw new Error(errorData.detail || 'Failed to process request');
      }
      
      const result = await response.json();
      await this.updateRequestStatus(requestData.request_id, 'pending');
      
      return result;
    } catch (error) {
      console.error('Request processing error:', error);
      if (requestData.request_id) {
        await this.updateRequestStatus(requestData.request_id, 'failed', {
          error: error instanceof Error ? error.message : 'Unknown error'
        });
      }
      throw error;
    }
  }

  // Status tracking
  private async updateRequestStatus(requestId: string, status: string, metadata?: any): Promise<void> {
    try {
      const { error } = await supabase
        .from('team_change_requests')
        .update({
          status,
          metadata: metadata ? { ...metadata, updated_at: new Date().toISOString() } : undefined
        })
        .eq('id', requestId);

      if (error) {
        console.error('Error updating request status:', error);
      }
    } catch (error) {
      console.error('Error updating request status:', error);
    }
  }
  
  async getRequest(requestId: string): Promise<any> {
    try {
      const response = await fetch(`${this.apiUrl}/requests/${requestId}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json'
        }
      });
      
      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(errorData.detail || 'Failed to get request');
      }
      
      return await response.json();
    } catch (error) {
      console.error('Error getting request:', error);
      return null;
    }
  }
  
  async getTeamRequests(teamId: string, status?: string): Promise<any[]> {
    try {
      // Use Supabase client for direct database access
      let query = supabase
        .from('team_change_requests')
        .select('*')
        .eq('team_id', teamId);
        
      if (status) {
        query = query.eq('status', status);
      }
      
      const { data, error } = await query.order('created_at', { ascending: false });
      
      if (error) {
        throw error;
      }
      
      return data || [];
    } catch (error) {
      console.error('Error getting team requests:', error);
      return [];
    }
  }
  // Method to fetch detailed tournament info
  async fetchTournamentDetails(tournamentId: string): Promise<any> {
    console.log(`Fetching details for tournament ID: ${tournamentId}`);
    try {
      // Assuming backend endpoint GET /api/tournaments/{tournamentId} exists
      const response = await fetch(`${this.apiUrl}/tournaments/${tournamentId}`);
      if (!response.ok) {
        // Attempt to parse error details from backend if available
        let errorDetails = `HTTP error! status: ${response.status}`;
        try {
          const errorData = await response.json();
          errorDetails = errorData.detail || errorDetails;
        } catch (e) { /* Ignore JSON parsing error */ }
        throw new Error(`Failed to fetch tournament details: ${errorDetails}`);
      }
      const data = await response.json();
      console.log("Tournament details fetched:", data);
      // TODO: Add validation here to ensure data matches expected structure
      return data;
    } catch (error) {
      console.error('Error in fetchTournamentDetails:', error);
      throw error; // Re-throw to be handled by the calling component
    }
  }

  // Method to fetch detailed league info
  async fetchLeagueDetails(leagueId: string): Promise<any> {
    console.log(`Fetching details for league ID: ${leagueId}`);
    try {
      // Assuming backend endpoint GET /api/leagues/{leagueId} exists
      const response = await fetch(`${this.apiUrl}/leagues/${leagueId}`);
      if (!response.ok) {
         // Attempt to parse error details from backend if available
        let errorDetails = `HTTP error! status: ${response.status}`;
        try {
          const errorData = await response.json();
          errorDetails = errorData.detail || errorDetails;
        } catch (e) { /* Ignore JSON parsing error */ }
        throw new Error(`Failed to fetch league details: ${errorDetails}`);
      }
      const data = await response.json();
      console.log("League details fetched:", data);
      // TODO: Add validation here to ensure data matches expected structure
      return data;
    } catch (error) {
      console.error('Error in fetchLeagueDetails:', error);
      throw error; // Re-throw to be handled by the calling component
    }
  }
}

File: backend/app/services/request_service.py
async def execute_approved_request(self, request_id: str) -> bool:
    """Execute an approved request after payment has been processed"""
    # Fetch the request from the database
    request_data = await self.get_request(request_id)
    
    if not request_data:
        logger.error(f"Request {request_id} not found")
        return False
        
    if request_data.get('status') != 'approved':
        logger.error(f"Cannot execute request {request_id} with status {request_data.get('status')}")
        return False
    
    # Execute the request based on its type
    try:
        request_type = request_data.get('request_type')
        
        if request_type == 'team_transfer':
            return await self._execute_team_transfer(request_data)
        elif request_type == 'roster_change':
            return await self._execute_roster_change(request_data)
        elif request_type == 'tournament_registration':
            return await self._execute_tournament_registration(request_data)
        elif request_type == 'league_registration':
            return await self._execute_league_registration(request_data)
        elif request_type == 'team_rebrand':
            return await self._execute_team_rebrand(request_data)
        elif request_type == 'online_id_change':
            return await self._execute_online_id_change(request_data)
        elif request_type == 'team_creation':
            return await self._execute_team_creation(request_data)
        else:
            logger.error(f"Unknown request type: {request_type}")
            return False
    except Exception as e:
        logger.exception(f"Error executing request {request_id}: {str(e)}")
        # Update the request status to failed
        await self._update_request_status(request_id, 'failed', {'error': str(e)})
        return False 

async def _execute_team_transfer(self, request_data: dict) -> bool:
    """Execute a team transfer request"""
    team_id = request_data.get('team_id')
    old_captain_id = request_data.get('old_captain_id')
    new_captain_id = request_data.get('new_captain_id')
    
    if not all([team_id, old_captain_id, new_captain_id]):
        logger.error(f"Missing required data for team transfer: {request_data}")
        return False
    
    try:
        # Update the team's captain_id
        result = await self.supabase.table('teams').update({
            'captain_id': new_captain_id,
            'updated_at': datetime.now().isoformat()
        }).eq('id', team_id).execute()
        
        if not result.data:
            logger.error(f"Failed to update team captain: {result.error}")
            return False
        
        # Update the request status to completed
        await self._update_request_status(request_data.get('request_id'), 'completed')
        return True
    except Exception as e:
        logger.exception(f"Error executing team transfer: {str(e)}")
        return False 

async def _update_request_status(self, request_id: str, status: str, metadata: dict = None) -> bool:
    """Update the status of a request"""
    try:
        update_data = {
            'status': status,
            'updated_at': datetime.now().isoformat()
        }
        
        if metadata:
            # If we have existing metadata, merge it
            request = await self.get_request(request_id)
            existing_metadata = request.get('metadata', {}) if request else {}
            
            if existing_metadata and isinstance(existing_metadata, dict):
                existing_metadata.update(metadata)
                update_data['metadata'] = existing_metadata
            else:
                update_data['metadata'] = metadata
        
        result = await self.supabase.table('team_change_requests').update(update_data).eq('request_id', request_id).execute()
        
        if not result.data:
            logger.error(f"Failed to update request status: {result.error}")
            return False
            
        return True
    except Exception as e:
        logger.exception(f"Error updating request status: {str(e)}")
        return False 

async def _execute_team_rebrand(self, request_data: dict) -> bool:
    """Execute a team rebrand request"""
    team_id = request_data.get('team_id')
    new_name = request_data.get('new_name')
    
    if not all([team_id, new_name]):
        logger.error(f"Missing required data for team rebrand: {request_data}")
        return False
    
    try:
        # Update the team's name
        result = await self.supabase.table('teams').update({
            'name': new_name,
            'updated_at': datetime.now().isoformat()
        }).eq('id', team_id).execute()
        
        if not result.data:
            logger.error(f"Failed to update team name: {result.error}")
            return False
        
        # Update the request status to completed
        await self._update_request_status(request_data.get('request_id'), 'completed')
        return True
    except Exception as e:
        logger.exception(f"Error executing team rebrand: {str(e)}")
        return False

async def _execute_online_id_change(self, request_data: dict) -> bool:
    """Execute an online ID change request"""
    team_id = request_data.get('team_id')
    player_id = request_data.get('player_id')
    new_online_id = request_data.get('new_online_id')
    platform = request_data.get('platform', 'psn')
    
    if not all([team_id, player_id, new_online_id]):
        logger.error(f"Missing required data for online ID change: {request_data}")
        return False
    
    try:
        # Update the online ID in league_roster
        league_result = await self.supabase.table('league_roster').update({
            'online_id': new_online_id,
            'updated_at': datetime.now().isoformat()
        }).eq('team_id', team_id).eq('player_id', player_id).execute()
        
        # Update the online ID in tournament_roster
        tournament_result = await self.supabase.table('tournament_roster').update({
            'online_id': new_online_id,
            'updated_at': datetime.now().isoformat()
        }).eq('team_id', team_id).eq('player_id', player_id).execute()
        
        # Update the request status to completed (even if no rows were updated)
        await self._update_request_status(request_data.get('request_id'), 'completed')
        return True
    except Exception as e:
        logger.exception(f"Error executing online ID change: {str(e)}")
        return False

async def _execute_roster_change(self, request_data: dict) -> bool:
    """Execute a roster change request"""
    team_id = request_data.get('team_id')
    player_id = request_data.get('player_id')
    operation = request_data.get('operation')
    new_role = request_data.get('new_role')
    
    if not all([team_id, player_id, operation]):
        logger.error(f"Missing required data for roster change: {request_data}")
        return False
    
    try:
        if operation == 'add':
            # Add player to team_players
            team_player_result = await self.supabase.table('team_players').insert({
                'team_id': team_id,
                'user_id': player_id,
                'role': new_role or 'player',
                'created_at': datetime.now().isoformat(),
                'updated_at': datetime.now().isoformat()
            }).execute()
            
            if not team_player_result.data:
                logger.error(f"Failed to add player to team: {team_player_result.error}")
                return False
                
        elif operation == 'remove':
            # Remove player from team_players
            team_player_result = await self.supabase.table('team_players').delete().eq('team_id', team_id).eq('user_id', player_id).execute()
            
            # Also remove from league_roster and tournament_roster
            await self.supabase.table('league_roster').delete().eq('team_id', team_id).eq('player_id', player_id).execute()
            await self.supabase.table('tournament_roster').delete().eq('team_id', team_id).eq('player_id', player_id).execute()
            
        elif operation == 'update':
            # Update player role
            team_player_result = await self.supabase.table('team_players').update({
                'role': new_role,
                'updated_at': datetime.now().isoformat()
            }).eq('team_id', team_id).eq('user_id', player_id).execute()
            
            if not team_player_result.data:
                logger.error(f"Failed to update player role: {team_player_result.error}")
                return False
        
        # Update the request status to completed
        await self._update_request_status(request_data.get('request_id'), 'completed')
        return True
    except Exception as e:
        logger.exception(f"Error executing roster change: {str(e)}")
        return False 

async def _execute_league_registration(self, request_data: dict) -> bool:
    """Execute a league registration request"""
    team_id = request_data.get('team_id')
    league_id = request_data.get('league_id')
    player_ids = request_data.get('player_ids', [])
    
    if not all([team_id, league_id]) or not player_ids:
        logger.error(f"Missing required data for league registration: {request_data}")
        return False
    
    try:
        # Register the team for the league
        for player_id in player_ids:
            # Check if player is already registered
            existing = await self.supabase.table('league_roster').select('*').eq('team_id', team_id).eq('league_id', league_id).eq('player_id', player_id).execute()
            
            if existing.data:
                # Update existing entry
                await self.supabase.table('league_roster').update({
                    'status': 'active',
                    'updated_at': datetime.now().isoformat()
                }).eq('team_id', team_id).eq('league_id', league_id).eq('player_id', player_id).execute()
            else:
                # Create new entry
                await self.supabase.table('league_roster').insert({
                    'team_id': team_id,
                    'league_id': league_id,
                    'player_id': player_id,
                    'status': 'active',
                    'created_at': datetime.now().isoformat(),
                    'updated_at': datetime.now().isoformat()
                }).execute()
        
        # Update the request status to completed
        await self._update_request_status(request_data.get('request_id'), 'completed')
        return True
    except Exception as e:
        logger.exception(f"Error executing league registration: {str(e)}")
        return False

async def _execute_tournament_registration(self, request_data: dict) -> bool:
    """Execute a tournament registration request"""
    team_id = request_data.get('team_id')
    tournament_id = request_data.get('tournament_id')
    player_ids = request_data.get('player_ids', [])
    
    if not all([team_id, tournament_id]) or not player_ids:
        logger.error(f"Missing required data for tournament registration: {request_data}")
        return False
    
    try:
        # Register the team for the tournament
        for player_id in player_ids:
            # Check if player is already registered
            existing = await self.supabase.table('tournament_roster').select('*').eq('team_id', team_id).eq('tournament_id', tournament_id).eq('player_id', player_id).execute()
            
            if existing.data:
                # Update existing entry
                await self.supabase.table('tournament_roster').update({
                    'status': 'active',
                    'updated_at': datetime.now().isoformat()
                }).eq('team_id', team_id).eq('tournament_id', tournament_id).eq('player_id', player_id).execute()
            else:
                # Create new entry
                await self.supabase.table('tournament_roster').insert({
                    'team_id': team_id,
                    'tournament_id': tournament_id,
                    'player_id': player_id,
                    'status': 'active',
                    'created_at': datetime.now().isoformat(),
                    'updated_at': datetime.now().isoformat()
                }).execute()
        
        # Update the request status to completed
        await self._update_request_status(request_data.get('request_id'), 'completed')
        return True
    except Exception as e:
        logger.exception(f"Error executing tournament registration: {str(e)}")
        return False 

async def _execute_team_creation(self, request_data: dict) -> bool:
    """Execute a team creation request"""
    team_id = request_data.get('team_id')
    team_name = request_data.get('team_name')
    captain_id = request_data.get('captain_id')
    league_id = request_data.get('league_id')
    
    if not all([team_id, team_name, captain_id]):
        logger.error(f"Missing required data for team creation: {request_data}")
        return False
    
    try:
        # Create the team
        team_result = await self.supabase.table('teams').insert({
            'id': team_id,
            'name': team_name,
            'captain_id': captain_id,
            'league_id': league_id,
            'status': 'active',
            'created_at': datetime.now().isoformat(),
            'updated_at': datetime.now().isoformat()
        }).execute()
        
        if not team_result.data:
            logger.error(f"Failed to create team: {team_result.error}")
            return False
        
        # Add captain to team_players
        player_result = await self.supabase.table('team_players').insert({
            'team_id': team_id,
            'user_id': captain_id,
            'role': 'captain',
            'created_at': datetime.now().isoformat(),
            'updated_at': datetime.now().isoformat()
        }).execute()
        
        if not player_result.data:
            logger.error(f"Failed to add captain to team: {player_result.error}")
            # Don't return false here, we want to continue
        
        # Update the request status to completed
        await self._update_request_status(request_data.get('request_id'), 'completed')
        return True
    except Exception as e:
        logger.exception(f"Error executing team creation: {str(e)}")
        return False 